File: Agent Library Dev Tucker/pantheon-go/kernel/value.go
package kernel

import (
	"encoding/json"
	"fmt"
)

// Kind is the discriminator of the Value union. Closed set of five, per spec §5.
// Adding a variant here is a change to the kernel and therefore a change to the
// Rust executor -- which is exactly the coupling invariant 1 forbids growing.
// If a new business concept seems to need a variant, it is a Record.
type Kind string

const (
	KindText   Kind = "text"
	KindFile   Kind = "file"
	KindTable  Kind = "table"
	KindRecord Kind = "record"
	KindError  Kind = "error"
)

// Text is a body of prose with an optional language tag.
type Text struct {
	Body string `json:"body"`
	Lang string `json:"lang,omitempty"`
}

// FileHandle is a content-addressed blob reference. Bodies never construct one:
// the hash is the executor's to compute and trust. Obtain handles from
// Ctx.PutBlob, which round-trips through the proxy (PROTOCOL.md, blob.put).
type FileHandle struct {
	Blob       string `json:"blob"`       // "sha256:..."
	MediaType  string `json:"media_type"` // spec §5: Bin folded into File, media type distinguishes
	Size       int64  `json:"size,omitempty"`
	Capability string `json:"capability,omitempty"`
	Filename   string `json:"filename,omitempty"`
}

// Column is one column's metadata in a TableHandle.
type Column struct {
	Name string `json:"name"`
	Type string `json:"type"` // "string" | "int" | "float" | "bool" | "timestamp"
}

// TableHandle points at a row source in the blob store: CSV or JSONL in v0,
// Arrow later (spec §11 defers it). Rows are read in chunks through the proxy
// so access stays metered and memory stays bounded.
type TableHandle struct {
	Blob       string   `json:"blob"`
	Format     string   `json:"format"` // "csv" | "jsonl"
	Columns    []Column `json:"columns"`
	Rows       int64    `json:"rows,omitempty"` // advisory; -1 or 0 when unknown
	Capability string   `json:"capability,omitempty"`
}

// Record carries all business meaning. The executor's entire relationship with
// it is validate(schema, data) -- which is what keeps the amount of Rust in the
// executor constant in the number of business types (invariant 1).
type Record struct {
	TypeRef TypeRef         `json:"type_ref"`
	Data    json.RawMessage `json:"data"`
}

// ErrorValue is a typed failure that flows as a value. A business failure is a
// Value the run log routes; a protocol fault is not (see PROTOCOL.md).
type ErrorValue struct {
	Code    string          `json:"code"`
	Message string          `json:"message"`
	Detail  json.RawMessage `json:"detail,omitempty"`
	Retry   bool            `json:"retry,omitempty"`
}

// Value is the kernel's tagged union, mirroring the Rust
// #[serde(tag = "kind", rename_all = "snake_case")] enum. serde's internally
// tagged representation flattens each variant's fields alongside "kind", so the
// custom marshalling below is not a stylistic choice -- it is what makes a Go
// runner and a Python runner produce byte-compatible envelopes.
type Value struct {
	Kind   Kind
	Text   *Text
	File   *FileHandle
	Table  *TableHandle
	Record *Record
	Error  *ErrorValue
}

// Constructors. Use these rather than building a Value literal: they keep Kind
// and the populated arm in agreement, which nothing else enforces.

// NewText builds a Text value with no language tag.
func NewText(body string) Value { return Value{Kind: KindText, Text: &Text{Body: body}} }

// NewFile wraps a content-addressed blob reference.
func NewFile(h FileHandle) Value { return Value{Kind: KindFile, File: &h} }

// NewTable wraps a row-source reference.
func NewTable(h TableHandle) Value { return Value{Kind: KindTable, Table: &h} }

// NewRecord marshals v as the record's data.
func NewRecord(ref TypeRef, v any) (Value, error) {
	b, err := json.Marshal(v)
	if err != nil {
		return Value{}, fmt.Errorf("kernel: marshalling record %s: %w", ref, err)
	}
	return Value{Kind: KindRecord, Record: &Record{TypeRef: ref, Data: b}}, nil
}

// NewRawRecord wraps already-marshalled data.
func NewRawRecord(ref TypeRef, data json.RawMessage) Value {
	return Value{Kind: KindRecord, Record: &Record{TypeRef: ref, Data: data}}
}

// NewError builds a typed failure value.
func NewError(code, message string) Value {
	return Value{Kind: KindError, Error: &ErrorValue{Code: code, Message: message}}
}

// IsZero reports whether the Value is unset.
func (v Value) IsZero() bool { return v.Kind == "" }

// Valuer lets an action's output type choose its own kernel representation.
// Without it every action output would be a Record; with it, an action can
// return a File or a Table as naturally as a struct. See action.Register.
type Valuer interface {
	KernelValue() Value
}

// UnmarshalData decodes a Record's data into dst. It is an error to call this
// on a non-Record, because silently succeeding on a Text would turn a wiring
// mistake into an empty struct three tasks downstream.
func (v Value) UnmarshalData(dst any) error {
	if v.Kind != KindRecord || v.Record == nil {
		return fmt.Errorf("kernel: cannot unmarshal data from a %q value; expected record", v.Kind)
	}
	return json.Unmarshal(v.Record.Data, dst)
}

// MarshalJSON emits the internally tagged form: the variant's fields sit
// alongside "kind", matching serde's representation on the Rust side.
func (v Value) MarshalJSON() ([]byte, error) {
	var (
		payload any
		err     error
	)
	switch v.Kind {
	case KindText:
		if v.Text == nil {
			return nil, errArmUnset(v.Kind)
		}
		payload = *v.Text
	case KindFile:
		if v.File == nil {
			return nil, errArmUnset(v.Kind)
		}
		payload = *v.File
	case KindTable:
		if v.Table == nil {
			return nil, errArmUnset(v.Kind)
		}
		payload = *v.Table
	case KindRecord:
		if v.Record == nil {
			return nil, errArmUnset(v.Kind)
		}
		payload = *v.Record
	case KindError:
		if v.Error == nil {
			return nil, errArmUnset(v.Kind)
		}
		payload = *v.Error
	case "":
		return nil, fmt.Errorf("kernel: cannot marshal a zero Value")
	default:
		return nil, fmt.Errorf("kernel: unknown value kind %q", v.Kind)
	}

	// Flatten the variant's fields alongside "kind", matching serde's
	// internally tagged representation.
	b, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}
	var m map[string]json.RawMessage
	if err := json.Unmarshal(b, &m); err != nil {
		return nil, err
	}
	m["kind"], _ = json.Marshal(string(v.Kind))
	return json.Marshal(m)
}

// UnmarshalJSON decodes the internally tagged form, refusing any kind outside
// the closed set of five.
func (v *Value) UnmarshalJSON(b []byte) error {
	var probe struct {
		Kind Kind `json:"kind"`
	}
	if err := json.Unmarshal(b, &probe); err != nil {
		return fmt.Errorf("kernel: value is not an object with a kind: %w", err)
	}
	out := Value{Kind: probe.Kind}
	switch probe.Kind {
	case KindText:
		out.Text = &Text{}
		if err := json.Unmarshal(b, out.Text); err != nil {
			return err
		}
	case KindFile:
		out.File = &FileHandle{}
		if err := json.Unmarshal(b, out.File); err != nil {
			return err
		}
	case KindTable:
		out.Table = &TableHandle{}
		if err := json.Unmarshal(b, out.Table); err != nil {
			return err
		}
	case KindRecord:
		out.Record = &Record{}
		if err := json.Unmarshal(b, out.Record); err != nil {
			return err
		}
	case KindError:
		out.Error = &ErrorValue{}
		if err := json.Unmarshal(b, out.Error); err != nil {
			return err
		}
	case "":
		return fmt.Errorf("kernel: value has no kind discriminator")
	default:
		// Refusing an unknown kind is deliberate. A Go runner that skipped a
		// variant it did not recognise would hand downstream tasks a zero
		// Value, and the failure would surface far from its cause.
		return fmt.Errorf("kernel: unknown value kind %q; the kernel is a closed set of five", probe.Kind)
	}
	*v = out
	return nil
}

func errArmUnset(k Kind) error {
	return fmt.Errorf("kernel: value of kind %q has no %s payload set; use the New%s constructor", k, k, k)
}

File: Agent Library Dev Tucker/pantheon-rs/crates/ptn-vocab/src/value.rs
//! The kernel `Value`: a closed set of five variants (spec §5).
//!
//! This is the vocabulary the executor has operational duties toward, and
//! nothing else. Store, hash and gate Files; meter Tables; validate Records
//! against their registered schema; route Errors. Everything with business
//! *meaning* is a `Record`, and the executor's entire relationship with one is
//! `validate(schema, data)`.
//!
//! Adding a variant here is a change to the executor, which is exactly the
//! coupling invariant 1 forbids growing. If a new business concept seems to
//! need a variant, it is a Record.

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::typeref::TypeRef;

/// A body of prose.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct Text {
    /// The prose itself.
    pub body: String,
    /// BCP 47 tag. Absent means unstated, not English.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub lang: Option<String>,
}

/// A content-addressed blob reference.
///
/// Bodies never construct one: the hash is the executor's to compute and trust.
/// Spec §5 folds Bin into File, so `media_type` is what distinguishes a PDF
/// from an HTML report from an mdzip — and understanding it is an action's job
/// at the edge, never the executor's.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct FileHandle {
    /// Content address, `sha256:<hex>`. The executor computes it; a body never does.
    pub blob: String,
    /// What the bytes are. Spec §5 folds Bin into File, so this is the only
    /// thing distinguishing a PDF from an HTML report.
    pub media_type: String,
    /// Bytes. Omitted when zero.
    #[serde(default, skip_serializing_if = "is_zero_u64")]
    pub size: u64,
    /// Grants the holder read access to this blob for the run's lifetime.
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub capability: String,
    /// Suggested name for a download. Advisory only.
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub filename: String,
}

/// One column's metadata in a table.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct Column {
    /// Column name. Rows are addressed by this, never by position.
    pub name: String,
    /// `string` | `int` | `float` | `bool` | `timestamp`.
    /// One of `string`, `int`, `float`, `bool`, `timestamp`.
    #[serde(rename = "type")]
    pub column_type: String,
}

/// Column metadata plus a blob-store row source.
///
/// CSV/JSONL in v0; Arrow is deferred (spec §11). Rows are read in chunks
/// through the proxy so access stays metered and memory stays bounded.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct TableHandle {
    /// Content address, `sha256:<hex>`. The executor computes it; a body never does.
    pub blob: String,
    /// `csv` or `jsonl` in v0. Arrow is deferred (spec §11).
    pub format: String,
    /// Declared, never inferred: inferring from row one makes a column of
    /// integers become a column of strings the day row one is null.
    pub columns: Vec<Column>,
    /// Advisory row count; zero when unknown.
    /// Advisory row count. Zero means unknown, not empty.
    #[serde(default, skip_serializing_if = "is_zero_u64")]
    pub rows: u64,
    /// Grants the holder read access to this blob for the run's lifetime.
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub capability: String,
}

/// A typed failure that flows as a value.
///
/// A business failure is a Value the run log routes; a protocol fault is not.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct ErrorValue {
    /// Stable, machine-readable failure class, e.g. `timeout`.
    pub code: String,
    /// Human-readable explanation.
    pub message: String,
    /// Structured context for whoever debugs this.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub detail: Option<serde_json::Value>,
    /// Whether retrying could plausibly succeed. Advisory to the executor.
    #[serde(default, skip_serializing_if = "is_false")]
    pub retry: bool,
}

/// The kernel's closed set of five (spec §5).
///
/// Internally tagged, so each variant's fields sit alongside `"kind"` in the
/// serialised object. That representation is the contract every runner
/// implementation shares; `tests/wire_compat.rs` pins it against a corpus the
/// Go runner reads too.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum Value {
    /// Prose. Human-facing output that is not a document.
    Text(Text),
    /// Stored bytes by content address. Any media type, including binary.
    File(FileHandle),
    /// Stored rows by handle, read in chunks so size stays bounded.
    Table(TableHandle),
    /// Carries all business meaning. The executor's entire relationship with
    /// one is `validate(schema, data)`, which is what keeps the amount of Rust
    /// in the executor constant in the number of business types (invariant 1).
    Record {
        /// Which registered schema `data` must validate against.
        type_ref: TypeRef,
        /// The business payload. Opaque to the executor.
        data: serde_json::Value,
    },
    /// A typed failure. Routed by the executor, not raised as a fault.
    Error(ErrorValue),
}

impl Value {
    /// The discriminator this value serialises with.
    pub fn kind(&self) -> &'static str {
        match self {
            Value::Text(_) => "text",
            Value::File(_) => "file",
            Value::Table(_) => "table",
            Value::Record { .. } => "record",
            Value::Error(_) => "error",
        }
    }

    /// A `Text` value with no language tag.
    pub fn text(body: impl Into<String>) -> Self {
        Value::Text(Text {
            body: body.into(),
            lang: None,
        })
    }

    /// A `Record` carrying already-serialised data.
    pub fn record(type_ref: TypeRef, data: serde_json::Value) -> Self {
        Value::Record { type_ref, data }
    }

    /// A non-retryable `Error` with no structured detail.
    pub fn error(code: impl Into<String>, message: impl Into<String>) -> Self {
        Value::Error(ErrorValue {
            code: code.into(),
            message: message.into(),
            detail: None,
            retry: false,
        })
    }

    /// The schema a produced value actually carries.
    ///
    /// An Error is a kernel value, not the task's declared output type.
    /// Labelling a failure with the success schema would make the run log lie.
    pub fn schema_ref(&self) -> TypeRef {
        match self {
            Value::Record { type_ref, .. } => type_ref.clone(),
            Value::Text(_) => TypeRef::new("kernel.Text", 1),
            Value::File(_) => TypeRef::new("kernel.File", 1),
            Value::Table(_) => TypeRef::new("kernel.Table", 1),
            Value::Error(_) => TypeRef::new("kernel.Error", 1),
        }
    }
}

fn is_zero_u64(v: &u64) -> bool {
    *v == 0
}

fn is_false(v: &bool) -> bool {
    !*v
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn is_internally_tagged() {
        let v = Value::text("hello");
        let got = serde_json::to_value(&v).unwrap();
        // Fields sit alongside "kind", not nested under it.
        assert_eq!(got, json!({"kind": "text", "body": "hello"}));
    }

    #[test]
    fn every_variant_round_trips() {
        let values = vec![
            Value::text("hi"),
            Value::File(FileHandle {
                blob: "sha256:ab".into(),
                media_type: "application/pdf".into(),
                size: 12,
                capability: String::new(),
                filename: String::new(),
            }),
            Value::Table(TableHandle {
                blob: "sha256:cd".into(),
                format: "csv".into(),
                columns: vec![Column {
                    name: "a".into(),
                    column_type: "int".into(),
                }],
                rows: 0,
                capability: String::new(),
            }),
            Value::record(TypeRef::new("Invoice", 2), json!({"total": 12.5})),
            Value::error("bad_input", "nope"),
        ];
        for v in values {
            let s = serde_json::to_string(&v).unwrap();
            let back: Value = serde_json::from_str(&s).unwrap();
            assert_eq!(v, back, "round trip changed {}", v.kind());
        }
    }

    /// The kernel is a closed set of five. A runner that skipped an
    /// unrecognised variant would hand downstream tasks nothing, and the
    /// failure would surface far from its cause.
    #[test]
    fn unknown_kind_is_refused() {
        let err = serde_json::from_str::<Value>(r#"{"kind":"quaternion","spin":2}"#);
        assert!(err.is_err());
    }

    #[test]
    fn missing_kind_is_refused() {
        assert!(serde_json::from_str::<Value>(r#"{"body":"hi"}"#).is_err());
    }

    #[test]
    fn empty_optionals_are_omitted() {
        // Matching the Go runner's omitempty rules is what keeps one corpus
        // valid for both implementations.
        let v = Value::File(FileHandle {
            blob: "sha256:ab".into(),
            media_type: "text/html".into(),
            size: 0,
            capability: String::new(),
            filename: String::new(),
        });
        let got = serde_json::to_value(&v).unwrap();
        assert_eq!(
            got,
            json!({"kind":"file","blob":"sha256:ab","media_type":"text/html"})
        );
    }

    #[test]
    fn an_error_is_labelled_as_an_error_not_as_the_success_type() {
        assert_eq!(Value::error("e", "m").schema_ref().name, "kernel.Error");
    }
}
